File: android/src/main/kotlin/live/hms/hmssdk_flutter/HMSMessageRecipientExtension.kt
package live.hms.hmssdk_flutter

import live.hms.video.sdk.models.HMSMessageRecipient
import live.hms.video.sdk.models.enums.HMSMessageRecipientType

class HMSMessageRecipientExtension {
    companion object {
        fun toDictionary(hmsMessageRecipient: HMSMessageRecipient?): HashMap<String, Any?>? {
            val hashMap = HashMap<String, Any?>()
            if (hmsMessageRecipient == null)return null

            hashMap["recipient_peer"] = HMSPeerExtension.toDictionary(hmsMessageRecipient.recipientPeer)

            val recipientRoles = ArrayList<HashMap<String, Any?>?>()
            hmsMessageRecipient.recipientRoles.forEach {
                recipientRoles.add(HMSRoleExtension.toDictionary(it))
            }
            hashMap["recipient_roles"] = if (recipientRoles.size != 0)recipientRoles else null

            hashMap["recipient_type"] = getValueOfHMSMessageRecipient(hmsMessageRecipient.recipientType)

            return hashMap
        }

        private fun getValueOfHMSMessageRecipient(hmsMessageRecipientType: HMSMessageRecipientType?): String? {
            if (hmsMessageRecipientType == null)return null
            return when (hmsMessageRecipientType) {
                HMSMessageRecipientType.BROADCAST -> "broadCast"
                HMSMessageRecipientType.PEER -> "peer"
                HMSMessageRecipientType.ROLES -> "roles"
                else -> "defaultRecipient"
            }
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Models/HMSMessageRecipientExtension.swift
//
//  HMSMessageRecipientExtension.swift
//  hmssdk_flutter
//
//  Created by Yogesh Singh on 30/09/21.
//

import Foundation
import HMSSDK

class HMSMessageRecipientExtension {

    static func toDictionary(_ receipient: HMSMessageRecipient) -> [String: Any] {

        var dict = [String: Any]()

        dict["recipient_type"] = getString(from: receipient.type)

        if let peer = receipient.peerRecipient {
            dict["recipient_peer"] = HMSPeerExtension.toDictionary(peer)
        }

        if let roles = receipient.rolesRecipient {
            var roleExtension = [Any]()
            roles.forEach { roleExtension.append(HMSRoleExtension.toDictionary($0)) }
            dict["recipient_roles"] = roleExtension
        }

        return dict
    }

    static func getString(from type: HMSMessageRecipientType) -> String {
        switch type {
        case .broadcast:
            return "broadCast"
        case .peer:
            return "peer"
        case .roles:
            return "roles"
        default:
            return "defaultRecipient"
        }
    }
}
